src/herder/test/HerderTests.cpp

```diff
@@ -1620,100 +1620,126 @@ TEST_CASE("SCP State", "[herder][acceptance]")
+    HerderImpl& BHerder = *static_cast<HerderImpl*>(&B->getHerder());
+    HerderImpl& CHerder = *static_cast<HerderImpl*>(&C->getHerder());
 
-    HerderImpl& herder = *static_cast<HerderImpl*>(&listener->getHerder());
-    HerderImpl& valHerder = *static_cast<HerderImpl*>(&validator->getHerder());
-    auto const& lm = listener->getLedgerManager();
-    auto const& cm = listener->getCatchupManager();
+    auto const& lm = C->getLedgerManager();
+    auto const& cm = C->getCatchupManager();
 
     // Now construct a few future externalize messages
     // Make sure out of order messages are still within the validity range
     std::map<uint32_t, std::pair<SCPEnvelope, TxSetFramePtr>>
-        validatorSCPMessages;
-    Hash qSetHash = sha256(xdr::xdr_to_opaque(qset));
+        AValidatorSCPMessages;
+    std::map<uint32_t, std::pair<SCPEnvelope, TxSetFramePtr>>
+        BValidatorSCPMessages;
 
-    // Advance validator a bit further, and collect its externalize messages
-    auto destinationLedger = waitForValidator(maxSlots / 2);
+    // Advance A and B a bit further, and collect externalize messages
+    auto destinationLedger = waitForA(4);
     for (auto start = currentLedger + 1; start <= destinationLedger; start++)
     {
-        auto envs = valHerder.getSCP().getLatestMessagesSend(start);
-        for (auto const& env : envs)
+        for (auto const& env : AHerder.getSCP().getLatestMessagesSend(start))
+        {
+            if (env.statement.pledges.type() == SCP_ST_EXTERNALIZE)
+            {
+                StellarValue sv;
+                auto& pe = AHerder.getPendingEnvelopes();
+                AHerder.getHerderSCPDriver().toStellarValue(
+                    env.statement.pledges.externalize().commit.value, sv);
+                auto txset = pe.getTxSet(sv.txSetHash);
+                REQUIRE(txset);
+                AValidatorSCPMessages[start] = std::make_pair(env, txset);
+            }
+        }
+
+        for (auto const& env : BHerder.getSCP().getLatestMessagesSend(start))
         {
             if (env.statement.pledges.type() == SCP_ST_EXTERNALIZE)
             {
                 StellarValue sv;
-                auto& pe = valHerder.getPendingEnvelopes();
-                valHerder.getHerderSCPDriver().toStellarValue(
+                auto& pe = BHerder.getPendingEnvelopes();
+                BHerder.getHerderSCPDriver().toStellarValue(
                     env.statement.pledges.externalize().commit.value, sv);
                 auto txset = pe.getTxSet(sv.txSetHash);
                 REQUIRE(txset);
-                validatorSCPMessages[start] = std::make_pair(env, txset);
+                BValidatorSCPMessages[start] = std::make_pair(env, txset);
             }
         }
     }
 
-    REQUIRE(herder.getCurrentLedgerSeq() == currentListenerLedger());
+    REQUIRE(AValidatorSCPMessages.size() == BValidatorSCPMessages.size());
+    REQUIRE(CHerder.getCurrentLedgerSeq() == currentCLedger());
```

Comment:
also need to re-check that ` REQUIRE(currentCLedger() == currentLedger);` in case there was a bug in the test (and C was still connected)